### DISCUSSION TITLE
TypeScript type mismatch with Joi (and testing plugins)

### DISCUSSION THREAD

**Discussion by webstackdev** [gatsbyjs/gatsby#35229] [RFC]
Since TS support for `gatsby-*` files landed in 4.9, I've been refactoring my core Gatsby files to take advantage of it. I've run into a problem involving types for the `pluginOptionsSchema` gatsby-node API method.

I think at one point the TS types for Joi (validation library) were out of sync as that project moved out of the @hapi namespace. The `gatsby-plugin-utils` has a [modified set of types](https://github.com/gatsbyjs/gatsby/blob/54d4721462b9303fed723fdcb15ac5d72e103778/packages/gatsby-plugin-utils/src/utils/plugin-options-schema-joi-type.ts) and [re-exports the Joi object](https://github.com/gatsbyjs/gatsby/blob/54d4721462b9303fed723fdcb15ac5d72e103778/packages/gatsby-plugin-utils/src/joi.ts), along with a `testPluginOptionsSchema` function.

I have some functionality that I implemented in a Joi extension so that it was easier to use template-d error messages. I need the Joi types for `CustomHelpers`, `Extension`, and`ExtensionFactory` which aren't exported from the hard-code version of Joi types in `gatsby-plugin-utils` (which removes the `Joi` namespace from the typings). The type is available on Joi e.g. `Joi.ExtensionFactory` from the `joi` package, but the two sets of typings (Joi's and `gatsby-plugin-utils`) have gotten out of sync. The Joi interface for `ValidationError` has an `_original: any` key, where the `gatsby-plugin-utils` version has an `_object: any` key instead so I get type errors.

My interim solution is a local copy of  `gatsby-plugin-utils` with modified typings. I think the right solution is to modify the Gatsby typings:

```typescript
import { Joi } from 'joi'
export interface PluginOptionsSchemaArgs {
  Joi: Joi.Root
}
```
Another possibility is to move [`packages/gatsby-plugin-util/src/joi.ts`](https://github.com/gatsbyjs/gatsby/blob/54d4721462b9303fed723fdcb15ac5d72e103778/packages/gatsby-plugin-utils/src/joi.ts) to either `packages/gatsby/src/bootstrap/load-plugins/` or `packages/gatsby/src/joi-schemas/` in an appropriate file, and modify the type of the exported Joi object to `Joi.Root`. That way there's a central place to manage typings and extensions to the Joi object used in plugin options validation.

I'd be happy to push a PR through but opened a discussion because I'm not sure the best route to go. There was some work on plugin option validation to use dot env vars in `gatsby-config`, but it doesn't seem like that work went anywhere and might be reasonably stripped out (note from @mxstbr October 2021 that the functionality was being considered). So I don't think the reason for providing a custom Gatsby-specific type for the Joi object used in plugin option schema validation may still even be relevant.

*On a tenuously related note:*

I plan to create a plugin with a few additional Joi validation testing methods in it. I'm not sure if it'll be useful to a wider audience, but it had me thinking about Gatsby testing plugins generally. I'll end up with a name like `gatsby-plugin-testing-joi` in a vendor namespace. There's lots of variations on the second slug segment for Gatsby plugins now; a `testing` one is a suggestion I'd add (e.g. `gatsby-testing-joi`). A Jest plugin seems like it could provide drop-in Jest config with all of the piddly files built in and reasonable defaults.

**Comment by moonmeister** [comment]
Hey @webstackdev, Some history:

In this [commit ](https://github.com/gatsbyjs/gatsby/pull/27381/commits/fdb65de11555b75605e593db8e0fe7e177c69bdd) I un-copied the custom types cause I didn't think we needed them. Max redid it in a later [commit](https://github.com/gatsbyjs/gatsby/pull/27381/commits/b20d463e7a84f21cf3d7743d0b972ec2360244c5). If I remember correctly it was something to do with Joi's typing for `external` async validations which we explicitly wanted to support. Ah, I found it, here's the [discussion](https://github.com/gatsbyjs/gatsby/pull/27381#discussion_r503131673) where we revert the change and the reason why.

I'm not sure the status of the `dotenv()` stuff you mentioned, but if we can find a way to not copy all the Joi types out to modify one thing then a PR would be great!
